File: app/src/main/java/com/raylabs/doggie/ui/categories/CategoriesViewModel.java
package com.raylabs.doggie.ui.categories;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.ViewModel;

import com.raylabs.doggie.data.DoggieRepository;
import com.raylabs.doggie.data.source.local.entity.DoggieEntity;
import com.raylabs.doggie.vo.Resource;

import java.util.List;

public class CategoriesViewModel extends ViewModel {

    private final DoggieRepository repository;
    private LiveData<Resource<List<DoggieEntity>>> data;

    public CategoriesViewModel(DoggieRepository repository) {
        this.repository = repository;
    }

    public LiveData<Resource<List<DoggieEntity>>> getData() {
        if (data == null) data = repository.getCategories();
        return data;
    }
}

File: app/src/main/java/com/raylabs/doggie/ui/categories/CategoriesViewModel.kt
package com.raylabs.doggie.ui.categories

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import androidx.paging.PagingData
import androidx.paging.cachedIn
import com.raylabs.doggie.data.DoggieRepository
import com.raylabs.doggie.vo.BreedCategory
import kotlinx.coroutines.flow.Flow

class CategoriesViewModel(
    private val repository: DoggieRepository
) : ViewModel() {

    val data: Flow<PagingData<BreedCategory>> =
        repository.getCategories().cachedIn(viewModelScope)

    fun requestPreview(category: BreedCategory) {
        repository.requestCategoryPreview(category)
    }
}
